File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.java
class Solution {
    public int longestIdealString(String s, int k) {
        int n = s.length();
        int[] dp = new int[26];
        int ans = 0;
        for (int i = 0; i < n; i++) {
            int c = s.charAt(i) - 'a';
            int best = 0;
            for (int j = Math.max(0, c - k); j < Math.min(26, c + k + 1); j++) {
                best = Math.max(best, dp[j]);
            }
            dp[c] = best + 1;
            ans = ans > dp[c] ? ans : dp[c];
        }
        return ans;
    }
}

File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.py
class Solution:
    def longestIdealString(self, s: str, k: int) -> int:
        nums = [ord(i) - ord('a') for i in s]
        n = len(nums)
        dp = [0] * 26
        for i in range(n):
            c = nums[i]
            best = 0
            for j in range(max(0, c - k), min(c + k + 1, 26)):
                best = max(best, dp[j])
            dp[c] = best + 1
        return max(dp)
                